File: NX python/SEQCMD/seqcmd_latency_parser.pl
#!/usr/bin/env perl

=head1 NAME

VODMA Command Latency Analyzer

=head1 DESCRIPTION

This script analyzes VODMA command latencies from ncverilog log files.

=head1 FEATURES

=over 4

=item * Tracks VODMA commands and their corresponding responses

=item * Calculates latency between command issue and completion

=item * Two analysis modes:
  - Separate mode: Analyze each VODMA type independently
  - Combined mode: Treat multiple VODMA types as a single client

=item * Provides detailed analysis including:
  - Command-by-command latency breakdown
  - Average latency across all commands
  - Maximum latency with detailed information
  - Total number of commands analyzed

=item * Time filtering: Analyze commands after specified timestamp (-t flag)

=back

=head1 USAGE

perl seqcmd_latency_parser.pl <logfile> [-c] [-t START_TIME] <vodma_type1> [vodma_type2 ...]

Options:
  -c : Combined mode - treat all VODMA types as single client
  -t START_TIME : Only analyze commands after this timestamp

Examples:
- Single type:     perl seqcmd_latency_parser.pl ncverilog.log vodma1yrr
- Separate types:  perl seqcmd_latency_parser.pl ncverilog.log vodma1yrr vodma1crr
- Combined types:  perl seqcmd_latency_parser.pl ncverilog.log -c vodma1yrr vodma1crr
- Time filter:     perl seqcmd_latency_parser.pl ncverilog.log -t 3000000 vodma1yrr

=cut

use strict;
use warnings;
use List::Util qw(sum max);
use Getopt::Long;
use Pod::Usage;

my $VERSION = '1.0';

# Main program
sub main {
    # Global variables
    my %vodma_commands; # tag -> [timestamp, vodma_type]
    my @latencies;      # [tag, vodma_type, start_time, end_time, latency]
    my @vodma_types;
    my %type_counts;    # Count of commands per type
    my $combined_mode = 0;
    my $start_time = 0;
    my $log_file;

    # Process command line arguments
    process_args(\$combined_mode, \$start_time, \$log_file, \@vodma_types);
    
    # Analyze the file
    analyze_file($log_file, $start_time, $combined_mode, \@vodma_types, \%vodma_commands, 
                 \@latencies, \%type_counts);
}

# Process command line arguments
sub process_args {
    my ($combined_mode_ref, $start_time_ref, $log_file_ref, $vodma_types_ref) = @_;
    my $help;
    
    GetOptions(
        "c" => $combined_mode_ref,
        "t=i" => $start_time_ref,
        "help|h" => \$help
    ) or pod2usage(1);
    
    pod2usage(-verbose => 2) if $help;
    
    # First argument should be log file
    $$log_file_ref = shift @ARGV;
    pod2usage("Error: No log file specified") unless defined $$log_file_ref;
    
    # Remaining arguments are vodma types
    @$vodma_types_ref = @ARGV;
    pod2usage("Error: No VODMA types specified") unless @$vodma_types_ref;
    
    # Print parsed arguments for verification
    print "\nParsed arguments:\n";
    print "Log file: $$log_file_ref\n";
    print "Start time: $$start_time_ref\n";
    print "VODMA types: ", join(", ", @$vodma_types_ref), "\n";
    print "Combined mode: ", $$combined_mode_ref ? "Yes" : "No", "\n\n";
}

# Parse a single line from the log file
sub parse_line {
    my ($line, $start_time, $vodma_types_ref, $vodma_commands_ref, 
        $latencies_ref, $type_counts_ref) = @_;
    
    # Extract timestamp at the start of the line
    my ($timestamp) = $line =~ /^(\d+)/;
    return unless defined $timestamp;
    
    # Skip if before start time
    return if $timestamp < $start_time;
    
    # Look for vodma commands
    foreach my $vodma_type (@$vodma_types_ref) {
        my ($tag) = $line =~ /\[seqcmd\].*\[\s*\Q$vodma_type\E\s*\].*tag\[(\d+)\]/;
        if (defined $tag) {
            $vodma_commands_ref->{$tag} = [$timestamp, $vodma_type];
            return;
        }
    }
    
    # Look for last data responses
    my ($tag) = $line =~ /\[data\].*tag\[(\d+)\].*last/;
    if (defined $tag && exists $vodma_commands_ref->{$tag}) {
        my ($cmd_start_time, $vodma_type) = @{$vodma_commands_ref->{$tag}};
        my $cmd_latency = $timestamp - $cmd_start_time;
        push @$latencies_ref, [$tag, $vodma_type, $cmd_start_time, $timestamp, $cmd_latency];
        $type_counts_ref->{$vodma_type}++;
        delete $vodma_commands_ref->{$tag};
    }
}

# Analyze the log file
sub analyze_file {
    my ($filename, $start_time, $combined_mode, $vodma_types_ref, 
        $vodma_commands_ref, $latencies_ref, $type_counts_ref) = @_;
    
    open(my $fh, '<', $filename) or die "Error: Cannot open file $filename: $!";
    
    while (my $line = <$fh>) {
        chomp $line;
        parse_line($line, $start_time, $vodma_types_ref, $vodma_commands_ref, 
                   $latencies_ref, $type_counts_ref);
    }
    
    close($fh);
    
    if ($combined_mode) {
        print_combined_analysis($start_time, $vodma_types_ref, $latencies_ref);
    } else {
        print_separate_analysis($start_time, $vodma_types_ref, $latencies_ref, $type_counts_ref);
    }
}

# Print analysis results in combined mode
sub print_combined_analysis {
    my ($start_time, $vodma_types_ref, $latencies_ref) = @_;
    
    if (@$latencies_ref == 0) {
        print "\nNo matching commands found with responses after timestamp $start_time.\n";
        return;
    }
    
    print "\nVodma Command Latency Analysis (Combined mode for: ", join(", ", @$vodma_types_ref), ")\n";
    print "Analysis starting from timestamp: $start_time\n";
    print "=" x 60, "\n";
    printf "%-10s %-12s %-12s %-8s\n", "Tag", "Start Time", "End Time", "Latency";
    print "-" x 60, "\n";
    
    # Sort latencies by start time and print
    foreach my $lat (sort { $a->[2] <=> $b->[2] } @$latencies_ref) {
        my ($tag, undef, $cmd_start, $cmd_end, $cmd_latency) = @$lat;
        printf "%-10s %-12s %-12s %-8s\n", $tag, $cmd_start, $cmd_end, $cmd_latency;
    }
    
    # Calculate stats
    my $total_latency = sum(map { $_->[4] } @$latencies_ref);
    my $avg_latency = $total_latency / scalar(@$latencies_ref);
    my $max_latency = max(map { $_->[4] } @$latencies_ref);
    my $max_lat_cmd;
    foreach my $lat (@$latencies_ref) {
        if ($lat->[4] == $max_latency) {
            $max_lat_cmd = $lat;
            last;
        }
    }
    
    print "\nSummary:\n";
    print "-" x 40, "\n";
    print "Total commands analyzed: ", scalar(@$latencies_ref), "\n";
    printf "Average latency: %.2f time units\n", $avg_latency;
    print "Maximum latency: $max_latency time units\n";
    
    print "\nMax latency command details:\n";
    print "  Tag: $max_lat_cmd->[0]\n";
    print "  Start time: $max_lat_cmd->[2]\n";
    print "  End time: $max_lat_cmd->[3]\n";
}

# Print analysis results in separate mode
sub print_separate_analysis {
    my ($start_time, $vodma_types_ref, $latencies_ref, $type_counts_ref) = @_;
    
    if (@$latencies_ref == 0) {
        print "\nNo matching commands found with responses after timestamp $start_time.\n";
        return;
    }
    
    print "\nVodma Command Latency Analysis for types: ", join(", ", @$vodma_types_ref), "\n";
    print "Analysis starting from timestamp: $start_time\n";
    print "=" x 80, "\n";
    printf "%-10s %-12s %-12s %-12s %-8s\n", "Tag", "Type", "Start Time", "End Time", "Latency";
    print "-" x 80, "\n";
    
    # Sort latencies by start time and print
    foreach my $lat (sort { $a->[2] <=> $b->[2] } @$latencies_ref) {
        my ($tag, $vtype, $cmd_start, $cmd_end, $cmd_latency) = @$lat;
        printf "%-10s %-12s %-12s %-12s %-8s\n", $tag, $vtype, $cmd_start, $cmd_end, $cmd_latency;
    }
    
    # Calculate overall stats
    my $total_latency = sum(map { $_->[4] } @$latencies_ref);
    my $avg_latency = $total_latency / scalar(@$latencies_ref);
    my $max_latency = max(map { $_->[4] } @$latencies_ref);
    my $max_lat_cmd;
    foreach my $lat (@$latencies_ref) {
        if ($lat->[4] == $max_latency) {
            $max_lat_cmd = $lat;
            last;
        }
    }
    
    print "\nSummary:\n";
    print "-" x 40, "\n";
    print "Total commands analyzed: ", scalar(@$latencies_ref), "\n";
    printf "Average latency: %.2f time units\n", $avg_latency;
    print "Maximum latency: $max_latency time units\n";
    
    print "\nBreakdown by type:\n";
    foreach my $vtype (sort keys %$type_counts_ref) {
        my @type_latencies = map { $_->[4] } grep { $_->[1] eq $vtype } @$latencies_ref;
        my $type_total = sum(@type_latencies);
        my $type_avg = $type_total / scalar(@type_latencies);
        print "  $vtype: $type_counts_ref->{$vtype} commands, avg latency: ";
        printf "%.2f\n", $type_avg;
    }
    
    print "\nMax latency command details:\n";
    print "  Tag: $max_lat_cmd->[0]\n";
    print "  Type: $max_lat_cmd->[1]\n";
    print "  Start time: $max_lat_cmd->[2]\n";
    print "  End time: $max_lat_cmd->[3]\n";
}

# Start execution
main();

File: NX python/SEQCMD/seqcmd_latency_parser.py
"""
VODMA Command Latency Analyzer
=============================

This script analyzes VODMA command latencies from ncverilog log files.

Features:
---------
- Tracks VODMA commands and their corresponding responses
- Calculates latency between command issue and completion
- Two analysis modes:
  * Separate mode: Analyze each VODMA type independently
  * Combined mode: Treat multiple VODMA types as a single client
- Provides detailed analysis including:
  * Command-by-command latency breakdown
  * Average latency across all commands
  * Maximum latency with detailed information
  * Total number of commands analyzed
  * Optional latency distribution plots (-p flag) saved to 'plots' directory
- Time filtering: Analyze commands after specified timestamp (-t flag)

Usage:
------
python seqcmd_parser.py <logfile> [-c] [-p] [-t START_TIME] <vodma_type1> [vodma_type2 ...]
Options:
  -c : Combined mode - treat all VODMA types as single client
  -p : Generate and save latency distribution plots to 'plots' directory
  -t START_TIME : Only analyze commands after this timestamp

Examples:
- Single type:     python seqcmd_parser.py ncverilog.log vodma1yrr
- Separate types:  python seqcmd_parser.py ncverilog.log vodma1yrr vodma1crr
- Combined types:  python seqcmd_parser.py ncverilog.log -c vodma1yrr vodma1crr
- With plots:      python seqcmd_parser.py ncverilog.log -p vodma1yrr
- Combined plots:  python seqcmd_parser.py ncverilog.log -c -p vodma1yrr vodma1crr
- Time filter:     python seqcmd_parser.py ncverilog.log -t 3000000 vodma1yrr
"""

import re
from typing import Dict, List, Tuple
import sys
from collections import defaultdict
import os
try:
    import matplotlib.pyplot as plt
    import numpy as np
    PLOT_AVAILABLE = True
except ImportError:
    PLOT_AVAILABLE = False

class VodmaLatencyAnalyzer:
    def __init__(self, vodma_types: List[str], combined_mode: bool = False, plot_enabled: bool = False, start_time: int = 0):
        self.vodma_commands: Dict[str, Tuple[int, str]] = {}  # tag -> (timestamp, vodma_type)
        self.latencies: List[Tuple[str, str, int, int, int]] = []  # (tag, vodma_type, start_time, end_time, latency)
        self.vodma_types = vodma_types
        self.type_counts = defaultdict(int)  # Count of commands per type
        self.combined_mode = combined_mode
        self.plot_enabled = plot_enabled and PLOT_AVAILABLE
        self.start_time = start_time
        
        # Create plots directory if plotting is enabled
        if self.plot_enabled:
            os.makedirs('plots', exist_ok=True)

    def save_plot(self, fig, filename):
        """Save plot to the plots directory"""
        filepath = os.path.join('plots', filename)
        fig.savefig(filepath)
        plt.close(fig)
        print(f"Plot saved to: {filepath}")

    def parse_line(self, line: str):
        # Extract timestamp at the start of the line
        timestamp_match = re.match(r'^(\d+)', line)
        if not timestamp_match:
            return
        
        timestamp = int(timestamp_match.group(1))
        
        # Skip if before start time
        if timestamp < self.start_time:
            return
        
        # Look for vodma commands
        for vodma_type in self.vodma_types:
            vodma_cmd_match = re.search(r'\[seqcmd\].*\[' + re.escape(vodma_type) + r'\s*\].*tag\[(\d+)\]', line)
            if vodma_cmd_match:
                tag = vodma_cmd_match.group(1)
                self.vodma_commands[tag] = (timestamp, vodma_type)
                return

        # Look for last data responses
        last_data_match = re.search(r'\[data\].*tag\[(\d+)\].*last', line)
        if last_data_match:
            tag = last_data_match.group(1)
            if tag in self.vodma_commands:
                start_time, vodma_type = self.vodma_commands[tag]
                latency = timestamp - start_time
                self.latencies.append((tag, vodma_type, start_time, timestamp, latency))
                self.type_counts[vodma_type] += 1
                del self.vodma_commands[tag]

    def plot_latency_distribution(self):
        if not self.plot_enabled:
            return

        # For combined mode
        if self.combined_mode:
            fig = plt.figure(figsize=(12, 6))
            latencies = [lat[4] for lat in self.latencies]
            
            # Histogram
            plt.subplot(121)
            plt.hist(latencies, bins=30, edgecolor='black')
            plt.title('Latency Distribution (All Types Combined)')
            plt.xlabel('Latency (time units)')
            plt.ylabel('Frequency')
            
            # Add mean and median lines
            mean_latency = np.mean(latencies)
            median_latency = np.median(latencies)
            plt.axvline(mean_latency, color='r', linestyle='dashed', linewidth=1, label=f'Mean: {mean_latency:.2f}')
            plt.axvline(median_latency, color='g', linestyle='dashed', linewidth=1, label=f'Median: {median_latency:.2f}')
            plt.legend()
            
            # Box plot
            plt.subplot(122)
            plt.boxplot(latencies, labels=['Combined'])
            plt.title('Latency Box Plot')
            plt.ylabel('Latency (time units)')
            
            plt.tight_layout()
            self.save_plot(fig, 'combined_latency_distribution.png')
        
        # For separate mode
        else:
            # Plot for each type
            for vtype in self.vodma_types:
                type_latencies = [lat[4] for lat in self.latencies if lat[1] == vtype]
                if not type_latencies:
                    continue
                
                fig = plt.figure(figsize=(12, 6))
                
                # Histogram
                plt.subplot(121)
                plt.hist(type_latencies, bins=30, edgecolor='black')
                plt.title(f'Latency Distribution - {vtype}')
                plt.xlabel('Latency (time units)')
                plt.ylabel('Frequency')
                
                # Add mean and median lines
                mean_latency = np.mean(type_latencies)
                median_latency = np.median(type_latencies)
                plt.axvline(mean_latency, color='r', linestyle='dashed', linewidth=1, label=f'Mean: {mean_latency:.2f}')
                plt.axvline(median_latency, color='g', linestyle='dashed', linewidth=1, label=f'Median: {median_latency:.2f}')
                plt.legend()
                
                # Box plot
                plt.subplot(122)
                plt.boxplot(type_latencies, labels=[vtype])
                plt.title('Latency Box Plot')
                plt.ylabel('Latency (time units)')
                
                plt.tight_layout()
                self.save_plot(fig, f'{vtype}_latency_distribution.png')
            
            # Comparison plot for all types
            if len(self.vodma_types) > 1:
                fig = plt.figure(figsize=(12, 6))
                data = []
                labels = []
                for vtype in self.vodma_types:
                    type_latencies = [lat[4] for lat in self.latencies if lat[1] == vtype]
                    if type_latencies:
                        data.append(type_latencies)
                        labels.append(vtype)
                
                plt.boxplot(data, labels=labels)
                plt.title('Latency Comparison Across Types')
                plt.ylabel('Latency (time units)')
                plt.xticks(rotation=45)
                plt.tight_layout()
                self.save_plot(fig, 'type_comparison.png')

    def analyze_file(self, filename: str):
        try:
            with open(filename, 'r') as f:
                for line in f:
                    self.parse_line(line.strip())
            
            if self.combined_mode:
                self._print_combined_analysis()
            else:
                self._print_separate_analysis()
            
            self.plot_latency_distribution()

        except FileNotFoundError:
            print(f"Error: File {filename} not found.")
        except Exception as e:
            print(f"Error processing file: {str(e)}")

    def _print_combined_analysis(self):
        if not self.latencies:
            print(f"\nNo matching commands found with responses after timestamp {self.start_time}.")
            return

        print(f"\nVodma Command Latency Analysis (Combined mode for: {', '.join(self.vodma_types)})")
        print(f"Analysis starting from timestamp: {self.start_time}")
        print("=" * 60)
        print(f"{'Tag':<10} {'Start Time':<12} {'End Time':<12} {'Latency':<8}")
        print("-" * 60)
        
        for tag, _, start, end, latency in sorted(self.latencies, key=lambda x: x[2]):
            print(f"{tag:<10} {start:<12} {end:<12} {latency:<8}")
        
        avg_latency = sum(lat[4] for lat in self.latencies) / len(self.latencies)
        max_latency = max(lat[4] for lat in self.latencies)
        max_lat_cmd = next(lat for lat in self.latencies if lat[4] == max_latency)
        
        print("\nSummary:")
        print("-" * 40)
        print(f"Total commands analyzed: {len(self.latencies)}")
        print(f"Average latency: {avg_latency:.2f} time units")
        print(f"Maximum latency: {max_latency} time units")
        
        print("\nMax latency command details:")
        print(f"  Tag: {max_lat_cmd[0]}")
        print(f"  Start time: {max_lat_cmd[2]}")
        print(f"  End time: {max_lat_cmd[3]}")

    def _print_separate_analysis(self):
        if not self.latencies:
            print(f"\nNo matching commands found with responses after timestamp {self.start_time}.")
            return

        print(f"\nVodma Command Latency Analysis for types: {', '.join(self.vodma_types)}")
        print(f"Analysis starting from timestamp: {self.start_time}")
        print("=" * 80)
        print(f"{'Tag':<10} {'Type':<12} {'Start Time':<12} {'End Time':<12} {'Latency':<8}")
        print("-" * 80)
        
        for tag, vtype, start, end, latency in sorted(self.latencies, key=lambda x: x[2]):
            print(f"{tag:<10} {vtype:<12} {start:<12} {end:<12} {latency:<8}")
        
        avg_latency = sum(lat[4] for lat in self.latencies) / len(self.latencies)
        max_latency = max(lat[4] for lat in self.latencies)
        max_lat_cmd = next(lat for lat in self.latencies if lat[4] == max_latency)
        
        print("\nSummary:")
        print("-" * 40)
        print(f"Total commands analyzed: {len(self.latencies)}")
        print(f"Average latency: {avg_latency:.2f} time units")
        print(f"Maximum latency: {max_latency} time units")
        
        print("\nBreakdown by type:")
        for vtype, count in self.type_counts.items():
            type_latencies = [lat[4] for lat in self.latencies if lat[1] == vtype]
            type_avg = sum(type_latencies) / len(type_latencies)
            print(f"  {vtype}: {count} commands, avg latency: {type_avg:.2f}")
        
        print("\nMax latency command details:")
        print(f"  Tag: {max_lat_cmd[0]}")
        print(f"  Type: {max_lat_cmd[1]}")
        print(f"  Start time: {max_lat_cmd[2]}")
        print(f"  End time: {max_lat_cmd[3]}")

def main():
    if len(sys.argv) < 3:
        print("Usage: python seqcmd_parser.py <logfile> [-c] [-p] [-t START_TIME] <vodma_type1> [vodma_type2 ...]")
        print("Options:")
        print("  -c : Combined mode - treat all VODMA types as single client")
        print("  -p : Generate and save latency distribution plots to 'plots' directory")
        print("  -t START_TIME : Only analyze commands after this timestamp")
        print("\nExamples:")
        print("  Single type:     python seqcmd_parser.py ncverilog.log vodma1yrr")
        print("  Separate types:  python seqcmd_parser.py ncverilog.log vodma1yrr vodma1crr")
        print("  Combined types:  python seqcmd_parser.py ncverilog.log -c vodma1yrr vodma1crr")
        print("  With plots:      python seqcmd_parser.py ncverilog.log -p vodma1yrr")
        print("  Combined plots:  python seqcmd_parser.py ncverilog.log -c -p vodma1yrr vodma1crr")
        print("  Time filter:     python seqcmd_parser.py ncverilog.log -t 3000000 vodma1yrr")
        print("  Order flexible:  python seqcmd_parser.py ncverilog.log vodma1yrr -t 3000000 -p")
        sys.exit(1)
    
    log_file = sys.argv[1]
    args = sys.argv[2:]
    
    # Initialize parameters
    combined_mode = False
    plot_enabled = False
    start_time = 0
    vodma_types = []
    
    # Process arguments
    i = 0
    while i < len(args):
        arg = args[i]
        if arg == '-c':
            combined_mode = True
            i += 1
        elif arg == '-p':
            plot_enabled = True
            i += 1
        elif arg == '-t':
            if i + 1 >= len(args):
                print("Error: -t option requires a timestamp value")
                sys.exit(1)
            try:
                start_time = int(args[i + 1])
            except ValueError:
                print(f"Error: Invalid timestamp value '{args[i + 1]}'. Must be an integer.")
                sys.exit(1)
            i += 2  # Skip both -t and its value
        elif arg.startswith('-'):
            print(f"Error: Unknown option '{arg}'")
            sys.exit(1)
        else:
            # This must be a vodma type
            vodma_types.append(arg)
            i += 1
    
    if not vodma_types:
        print("Error: No VODMA types specified")
        sys.exit(1)
    
    if plot_enabled and not PLOT_AVAILABLE:
        print("Warning: Plotting requires matplotlib and numpy. Please install them first.")
        print("You can install them using: pip install matplotlib numpy")
        plot_enabled = False
    
    # Print parsed arguments for verification
    print("\nParsed arguments:")
    print(f"Log file: {log_file}")
    print(f"Start time: {start_time}")
    print(f"VODMA types: {', '.join(vodma_types)}")
    print(f"Combined mode: {combined_mode}")
    print(f"Plot enabled: {plot_enabled}\n")
    
    analyzer = VodmaLatencyAnalyzer(vodma_types, combined_mode, plot_enabled, start_time)
    analyzer.analyze_file(log_file)

if __name__ == "__main__":
    main()
